toolchain/check/eval.cpp

```diff
@@ -1893,10 +1894,46 @@ static auto TryEvalInstInContext(EvalContext& eval_context,
       return MakeConstantResult(eval_context.context(), bind, phase);
     }
 
-    // These semantic wrappers don't change the constant value.
+    // AsCompatible changes the type of the source instruction; its constant
+    // value, if there is one, needs to be modified to be of the same type.
     case CARBON_KIND(SemIR::AsCompatible inst): {
-      return eval_context.GetConstantValue(inst.source_id);
+      auto value = eval_context.GetConstantValue(inst.source_id);
+      if (!value.is_constant()) {
+        return value;
+      }
+
+      auto type_id = inst.type_id;
+      auto value_inst = eval_context.insts().Get(
+          eval_context.constant_values().GetInstId(value));
+      value_inst.SetType(type_id);
+
+      auto from_phase = value.is_symbolic() ? Phase::Symbolic : Phase::Template;
+      auto to_phase = eval_context.types().GetConstantId(type_id).is_symbolic()
+                          ? Phase::Symbolic
+                          : Phase::Template;
+      if (to_phase >= from_phase) {
+        // If moving from a template constant value to a symbolic type, the new
+        // constant value takes on the phase of the new type. We're adding the
+        // symbolic bit to the new constant value due to the presence of a
+        // symbolic type.
+        return MakeConstantResult(eval_context.context(), value_inst, to_phase);
+      } else {
+        // If moving from a symbolic constant value to a template type, the new
+        // constant value has a phase that depends on what is in the value. If
+        // there is anything symbolic within the value, then it's symbolic. We
+        // can't easily determine that here without evaluating a new constant
+        // value. See
+        // https://github.com/carbon-language/carbon-lang/pull/4881#discussion_r1939961372
+        //
+        // TODO: Find a way to avoid recursion for this. Make this function into
+        // a while loop? Track the phase of the constant value separately from
+        // the phase of the value's type?
+        return TryEvalInstInContext(eval_context, SemIR::InstId::None,
```

Comment:
Can you use `[[clang::musttail]]` here? (I'm not sure if that suppresses the lint check, but if not we can file a clang-tidy bug...)
```suggestion
        [[clang::musttail]] return TryEvalInstInContext(eval_context, SemIR::InstId::None,
```

Reply:
https://godbolt.org/z/1WebxfofP

Looks like it does not suppress it. Filed https://github.com/llvm/llvm-project/issues/125724